### PR TITLE
add managed local kek with service against current proto api

### LINKED ISSUE #15: Add dedicated proto buf for ciphertext
## What

Add a protobuf api for encrypted blob that contains ciphertext and nonce.

## Why

Currently the nonce is prepending the ciphertext. This is limiting as the nonce size can't be change. Ever.

### PR DESCRIPTION
**Pull request by ibihim**
- [ ] Add proto that contains ciphertext and nonce (don't prepend ciphertext with nonce).

fixes: #15

### CODE REVIEW

File: kms/managed.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+)
+
+var (
+	// ErrKeyExpired means that the expiration time of a key has come and it shouldn't be used any more.
+	ErrKeyExpired = errors.New("key is out of date and shouldn't be used anymore for encryption")
+	// ErrNoCipher means that there is no upstream kms given and therefore the keys in use can't be protected.
+	ErrNoCipher = errors.New("no upstream encryption service was specified")
+
+	// expirationTime of a Week
+	expirationTime = time.Hour * 24 * 7
```
Comment (ritazh) [inline kms/managed.go]: if it's hardcoded to a week, then the var name should reflect that. e.g. `expirationTimeOneWeek`

File: kms/managed.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+		return nil, err
+	}
+
+	klog.Infof("new managed cipher is created")
+
+	return &mk, nil
+}
+
+func (m *ManagedCipher) manageKey() error {
+	// Expensive lock that stops the world until we have a new key set up.
```
Comment (ibihim) [inline kms/managed.go]: A better solution would be to prepare already the next cipher.

File: kms/managed.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		klog.Infof(
+			"decrypt key (%q) by upstream:",
+			base64.StdEncoding.EncodeToString(encKey),
+			err,
+		)
+
+		return nil, err
+	}
+
+	// Assume that keyID is the most up to date upstream KMS key.
```
Comment (ibihim) [inline kms/managed.go]: Yes, this one is tricky. Do we assume that new is always better? Is there a better strategy?

Comment (ibihim) [reply kms/managed.go]: Removed after hint from @aramase

**Comment by ibihim** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/112898

**Pull request closed without merging**